README.md
[![Travis Build Status](https://travis-ci.org/tuxette/RNAseqNet.svg?branch=master)](https://travis-ci.org/tuxette/RNAseqNet)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/tuxette/rnaseqnet?branch=master&svg=true)](https://ci.appveyor.com/project/tuxette/rnaseqnet)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/RNAseqNet)](https://CRAN.R-project.org/package=RNAseqNet)


# RNAseqNet

Infer log-linear Poisson Graphical Model with an auxiliary dataset. Hot-deck
multiple imputation method is used to improve the reliability of the inference
ith an auxiliary dataset. Standard log-linear Poisson GM can also be used for 
the inference and the StARS criterion is implemented to drive the selection of
the regularization parameter.

See Citation file (in `./inst/CITATION`) for citation details and the User's 
Guide included in `./inst/misc/RNAseqNetUsersGuide.html` for an example of 
usage.


NEWS
Version 0.1.4 [2020-06-08]
  bug fixes
    removed S3 class exportation

*******************************************************************************
Version 0.1.3 [2020-03-04]
  bug fixes
    fixed a bug related to a change in stringsAsFactor policy

  misc
    updated authors' names and emails

*******************************************************************************
Version 0.1.2
  remove Biocstyle for vignette formatting (instable links with Bioconductor)
  updated references with published article

*******************************************************************************
Version 0.1.1 [2017-05-16]
  new features
    GLMnetToGraph can now be used with matrices

  improved
    column names of the count data are now included in the result of GLMnetwork

  bug fixes
    fixed two bugs in plot.HDpath

  documentation
    added a vignette and a complete user guide
    fixed a few typos
    fixed an imprecise statement in print.GLMpath documentation
    updated references
    improved README file

*******************************************************************************
Version 0.1.0 [2017-04-19]
  Initial release


R/data.R

##########################
## Lung data (X)
#############################
#' @description
#' This data set is a small subset of the full data set from GTEx.  It contains
#' RNA-seq expressions measured from lung tissue. The RNA-seq expressions have
#' been normalized with the TMM method.
#'
#' @name lung
#' @title RNA-seq expression from lung tissue (GTEx).
#' @docType data
#' @format a data frame with 221 rows and 100 variables (genes). Rownames are
#' identifiers for individuals.
#'
#' @source The raw data were download from \url{https://gtexportal.org/}.
#' The TMM normalisation of RNA-seq expression was performed with the R package
#' \code{edgeR}.
#'
#' @author Alyssa Imbert <alyssa.imbert@gmail.com>

NULL


##########################
## Thyroid data (Y)
#############################
#' @description
#' This data set is a small subset of the full data set from GTEx.  It contains
#' RNA-seq expressions measured from thyroid tissue. The RNA-seq expressions
#' have been normalized with the TMM method.
#'
#' @docType data
#' @title RNA-seq expression from thyroid tissue (GTEx).
#' @name thyroid
#' @format a data frame with 221 rows and 50 variables (genes). Rownames are
#' identifiers for individuals.
#'
#' @source The raw data were downloaded from \url{https://gtexportal.org/}. The
#' TMM normalisation of RNA-seq expression was performed with the R package
#' \code{edgeR}.
#' @author Alyssa Imbert <alyssa.imbert@gmail.com>

NULL


R/donors.R
# manage donors and imputation
#
################################################################################
# donor pools, selection of sigma, imputed datasets
################################################################################
#' @import hot.deck
#' @importFrom hot.deck hot.deck
#'
#' @title Impute missing row datasets with multiple hot deck.
#' @export
#'
#' @description
#' \code{imputeHD} performs multiple hot-deck imputation on an input data frame
#' with missing rows. Each missing row is imputed with a unique donor. This
#' method requires an auxiliary dataset to compute similaritities between
#' individuals and create the pool of donors.
#'
#' @param X n x p numeric matrix containing RNA-seq expression with missing rows
#' (numeric matrix or data frame)
#' @param Y auxiliary dataset (n' x q numeric matrix or data frame)
#' @param sigma threshold for hot-deck imputation (numeric, positive)
#' @param m number of replicates in multiple imputation (integer). Default to 50
#' @param seed single value, interpreted as an in integer, used to initialize
#' the random number generation state. Default to \code{NULL} (not used in this
#' case)
#'
#' @details Missing values are identified by matching rownames in \code{X} and
#' \code{Y}. If rownames are not provided the missing rows in \code{X} are
#' supposed to correspond to the last rows of \code{Y}.
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr} }
#'
#' @references {Imbert, A., Valsesia, A., Le Gall, C., Armenise, C., Lefebvre,
#' G. Gourraud, P.A., Viguerie, N. and Villa-Vialaneix, N. (2018) Multiple
#' hot-deck imputation for network inference from RNA sequencing data.
#' \emph{Bioinformatics}. \doi{10.1093/bioinformatics/btx819}.}
#'
#' @seealso \code{\link{chooseSigma}}, \code{\link{imputedGLMnetwork}}
#'
#' @examples
#' data(lung)
#' data(thyroid)
#' nobs <- nrow(lung)
#' miss_ind <- sample(1:nobs, round(0.2 * nobs), replace = FALSE)
#' lung[miss_ind, ] <- NA
#' lung <- na.omit(lung)
#' imputed_lung <- imputeHD(lung, thyroid, sigma = 2)
#'
#' @return S3 object of class \code{HDImputed}: a list consisting of
#' \itemize{
#'   \item{\code{donors}}{ a list. Each element of this list contains the donor
#'   pool for every missing observations}
#'   \item{\code{draws}}{ a data frame which indicates which donor was chosen
#'   for each missing samples}
#'   \item{\code{data}}{ a list of \code{m} imputed datasets}
#' }

imputeHD <- function(X, Y, sigma, m = 50, seed = NULL) {
  if (!is.null(seed)) set.seed(seed)

  # prepare data
  if (is.null(rownames(X)) | is.null(rownames(Y))) {
    warning("rownames are not provided: the first ", nrow(X),
            "rows of Y are supposed to correspond to rows of X")
    sample_miss <- c(1:nrow(X), rep(NA, nrow(Y) - nrow(X)))
    merged_data <- data.frame("samples" = sample_miss, Y,
                              stringsAsFactors = TRUE)
  } else {
    s_size <- length(union(rownames(X), rownames(Y)))
    sample_miss <- c(intersect(rownames(X), rownames(Y)),
                     setdiff(rownames(Y), rownames(X)))
    merged_data <- data.frame("samples" = sample_miss,
                              Y[match(sample_miss, rownames(Y)), ],
                              stringsAsFactors = TRUE)
    merged_data$samples[is.na(match(merged_data$samples, rownames(X)))] <- NA
  }

  # hot deck imputation
  suppressWarnings(hd_res <- hot.deck(merged_data, sdCutoff = sigma, m = m))
  imputation <- lapply(hd_res$draws, function(alist) match(alist, rownames(X)))
  imputation <- matrix(unlist(imputation), byrow = TRUE,
                       nrow = sum(is.na(merged_data$samples)))
  imputation <- data.frame(imputation)

  # imputed datasets
  X_MI <- sapply(imputation, function(acol) {
    res <- rbind(X, X[acol, ])
    return(res)
  }, simplify = FALSE)

  # donors
  donors <- lapply(hd_res$donors, function(alist) match(alist, rownames(X)))

  res <- list("donors" = donors, "draws" = imputation, "data" = X_MI)
  class(res) <- "HDImputed"
  return(res)
}

## Methods for objects of class HDImputed
#' @title Methods for 'HDImputed' objects.
#' @name HDImputed
#' @export
#'
#' @aliases summary.HDImputed
#' @aliases print.HDImputed
#' @aliases HDImputed-class
#'
#' @description Methods for the result of \code{\link{imputeHD}}
#' (\code{HDImputed} object)
#' @param object \code{HDImputed} object
#' @param x \code{HDImputed} object
#' @param ... not used
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @seealso \code{\link{imputeHD}}

summary.HDimputed <- function(object, ...) {
  print(object)
  pool_len <- sapply(object$donors, length)
  cat("\nThe number of donors per missing sample is:\n")
  print(summary(pool_len))
  unique_draws <- apply(object$draws, 2, function(acol) length(unique(acol)))
  cat("The number of unique imputations per sample is:\n")
  print(summary(unique_draws))
}

#' @export
#' @rdname HDImputed
print.HDImputed <- function(x, ...) {
  cat(ncol(x$draws), "imputed datasets for", nrow(x$draws), "missing samples:\n",
      "donors are in '$donors' (list)\n",
      "chosen individuals are in '$draws' (matrix)\n",
      "imputed datasets are in '$data' (list)")
}

createDonors <- function(X, Y, sigma, seed = NULL) {
  if (!is.null(seed)) set.seed(seed)

  # prepare data
  if (is.null(rownames(X)) | is.null(rownames(Y))) {
    warning("rownames are not provided: the first ", nrow(X),
            "rows of Y are supposed to correspond to rows of X")
    sample_miss <- c(1:nrow(X), rep(NA, nrow(Y) - nrow(X)))
    merged_data <- data.frame("samples" = sample_miss, Y,
                              stringsAsFactors = TRUE)
  } else {
    s_size <- length(union(rownames(X), rownames(Y)))
    sample_miss <- c(intersect(rownames(X), rownames(Y)),
                     setdiff(rownames(Y), rownames(X)))
    merged_data <- data.frame("samples" = sample_miss,
                              Y[match(sample_miss, rownames(Y)), ],
                              stringsAsFactors = TRUE)
    merged_data$samples[is.na(match(merged_data$samples, rownames(X)))] <- NA
  }

  # hot deck imputation
  suppressWarnings(hd_res <- hot.deck(merged_data, sdCutoff = sigma, m = 1))
  donors <- lapply(hd_res$donors, function(alist) match(alist, rownames(X)))
  return(donors)
}

################################################################################
# variability of a donor pool
################################################################################
#' @importFrom stats dist
#' @title Average intra-donor pool variance.
#' @export
#'
#' @description
#' \code{varIntra} computes the average intra-donor pool variance.
#'
#' @param X n x p numeric matrix containing RNA-seq expression with missing rows
#' (numeric matrix or data frame)
#' @param Y auxiliary dataset (n' x q numeric matrix or data frame)
#' @param donors donor pool (a list, as given \code{$donors} obtained from the
#' function \code{\link{imputeHD}})
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @references {Imbert, A., Valsesia, A., Le Gall, C., Armenise, C., Lefebvre,
#' G. Gourraud, P.A., Viguerie, N. and Villa-Vialaneix, N. (2018) Multiple
#' hot-deck imputation for network inference from RNA sequencing data.
#' \emph{Bioinformatics}. \doi{10.1093/bioinformatics/btx819}.}
#'
#' @seealso \code{\link{imputeHD}}, \code{\link{chooseSigma}}
#'
#' @return \code{varIntra} returns a numeric value which is the average
#' intra-donor pool variance, as described in (Imbert \emph{et al.}, 2018).

varIntra <- function(X, Y, donors) {
  # prepare data
  if (is.null(rownames(X)) | is.null(rownames(Y))) {
    warning("rownames are not provided: the first ", nrow(X),
            "rows of Y are supposed to correspond to rows of X")
    indX <- 1:nrow(X)
    indY <- (nrow(X)+1):(nrow(Y))
  } else {
    indX <- intersect(rownames(X), rownames(Y))
    indY <- setdiff(rownames(Y), rownames(X))
  }

  all_varintra <- mapply(function(alist, ind) {
    all_dist <- apply(Y[indX, ][alist, , drop = FALSE], 1, function(arow)
      dist(rbind(arow, Y[ind, ]))^2
    )
    sum(all_dist) / length(all_dist)
  }, donors, indY)
  return(mean(all_varintra))
}

################################################################################
# diagnostic tool for choosing sigma
################################################################################
#'
#' @title Select the threshold sigma for hd-MI.
#' @export
#'
#' @description
#' \code{chooseSigma} computes the average intra-donor pool variance for
#' different values of sigma. It helps choosing a sigma that makes a good
#' trade-off between homogeneity within the pool of donors and variety (large
#' enough number of donors in every pool).
#'
#' @param X n x p numeric matrix containing RNA-seq expression with missing rows
#' (numeric matrix or data frame)
#' @param Y auxiliary dataset (n' x q numeric matrix or data frame)
#' @param sigma_list a sequence of increasing positive values for sigma (numeric
#' vector)
#' @param seed single value, interpreted as an in integer, used to initialize
#' the random number generation state
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @references {Imbert, A., Valsesia, A., Le Gall, C., Armenise, C., Lefebvre,
#' G. Gourraud, P.A., Viguerie, N. and Villa-Vialaneix, N. (2018) Multiple
#' hot-deck imputation for network inference from RNA sequencing data.
#' \emph{Bioinformatics}. \doi{10.1093/bioinformatics/btx819}.}
#'
#' @details The average intra-donor pool variance is described in (Imbert
#' \emph{et al.}, 2018).
#'
#' @seealso \code{\link{varIntra}}
#'
#' @examples
#' data(lung)
#' data(thyroid)
#' nobs <- nrow(lung)
#' miss_ind <- sample(1:nobs, round(0.2 * nobs), replace = FALSE)
#' lung[miss_ind, ] <- NA
#' lung <- na.omit(lung)
#' sigma_stats <- chooseSigma(lung, thyroid, 1:5)
#' \dontrun{plot(sigma_stats, type = "b")}
#'
#' @return a data frame with the values of sigma and the corresponding
#' intra-donor pool variances

chooseSigma <- function(X, Y, sigma_list, seed = NULL) {
  if (!is.null(seed)) set.seed(seed)

  donors <- lapply(sigma_list, function(sigma)
    createDonors(X, Y, sigma = sigma)
  )

  all_vars <- lapply(donors, function(alist) varIntra(X, Y, alist)) %>% unlist()

  res <- data.frame("sigma" = sigma_list, "varintra" = all_vars)
  return(res)
}


R/poissonnet.R
## TODO : documentation avec import de glmnet

# donor pools, selection of sigma, imputed datasets
################################################################################
#' @import glmnet
#' @import PoiClaClu
#' @importFrom glmnet glmnet
#' @importFrom PoiClaClu FindBestTransform
#'
#' @title Infer a network from RNA-seq expression
#' @export
#'
#' @description
#' \code{GLMnetwork} infers a network from RNA-seq expression by using a log-linear Poisson
#' graphical model 
#'
#' @param counts a n x p matrix of RNA-seq expression (numeric matrix or data frame)
#' @param lambdas a sequence of decresing positive numbers to control
#'  the regularization (numeric vector)
#' @param normalize logical value to normalize predictors in the log-linear
#' Poisson graphical model bu using a log transformation, default to TRUE
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@inra.fr}
#'
#' Nathalie Villa-Vialaneix, \email{nathalie.villa-vialaneix@inra.fr}}
#'
#' @seealso \code{\link{stabilitySelection}}
#' 
#'
#' @references{}
#' @return S3 object of class \code{GLMnetwork} ....
#'    

GLMnetwork <- function(counts, lambdas, normalize = TRUE) {
  nvar <- ncol(counts)
  nobs <- nrow(counts)
  # transformation of data (into Poisson distribution)
  alpha <- FindBestTransform(counts)
  tcounts <- counts^alpha

  # glm fitting for every variable
  all_glms <- lapply(1:ncol(counts), function(avar) {
    target <- counts[ ,avar,drop = FALSE]
    predictors <- counts[ ,-avar]
    if (normalize) {
      predictors <- log(predictors + 1)
      predictors <- scale(predictors, center = FALSE)
    }
    cur_glm <- glmnet(x = predictors, y = target, family = "poisson",
                      lambda = lambdas)
    return(cur_glm$beta)
  })

  # re-order result by lambda and fill the diagonal with 1s
  lambda_path <- lapply(seq_along(lambdas), function(indl) {
    all_betas <- lapply(all_glms, function(alist) alist[ ,indl]) %>% unlist()
    all_betas <- matrix(all_betas, byrow = TRUE, nrow = ncol(counts))
    res <- matrix(0, ncol = ncol(counts), nrow = ncol(counts))
    res[upper.tri(res)] <- all_betas[upper.tri(all_betas, diag = TRUE)]
    res[lower.tri(res)] <- all_betas[lower.tri(all_betas)]
    diag(res) <- 1
    return(res)
  })

  class(lambda_path) <- "GLMpath"
  return(lambda_path)
}

summary.GLMpath <- function(object, ...) {
  print(object)
}

#' @export
#' @rdname GLMpath
print.GLMpath <- function(x, ...) {
  cat("Object of class 'GLMpath'...\n",
      length(x), "predicted sets of coefficients from Poisson GLM.")
}

############################################################
## Selection of lambda
############################################################
#'
#' @import glmnet
#' @import PoiClaClu
#' @importFrom glmnet glmnet
#' @importFrom PoiClaClu FindBestTransform
#'
#' @title Selection of the regularization parameter for high dimensional undirected graph
#' @export
#'
#' @description
#' \code{stabilitySelection} implements the regularization parameter selection by using 
#' Stability Approach to Regularization Selection (StARS).
#'
#' @param counts a n x p matrix of RNA-seq expression (numeric matrix or data frame)
#' @param lambdas a sequence of decresing positive numbers to control
#'  the regularization (numeric vector)
#' @param B number of iterations for stability selection, default to 20
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@inra.fr}
#'
#' Nathalie Villa-Vialaneix, \email{nathalie.villa-vialaneix@inra.fr}}
#' 
#' @seealso \code{\link{GLMnetwork}}
#' 
#' @examples 
#' data(lung)
#' lambdas <- seq(5, 0.1, -0.1)
#' stab <- stabilitySelection(lung, lambdas)
#' plot(stab)
#'
#' @return S3 object of class \code{stabilitySelection} : a list consisiting of
#'  \itemize{
#'        \item{\code{lambdas}}{ numeric vector used for regularization path}
#'        \item{\code{B}}{number of iterations for stability selection}
#'        \item{\code{best}}{lambda value that gives the optimal network}
#'        \item{\code{variabilities}}{numeric vector ...}
#'    }
#'

stabilitySelection <- function(counts, lambdas, B = 20) {
  nvar <- ncol(counts)
  nobs <- nrow(counts)
  # transformation of data (into Poisson distribution)
  alpha <- FindBestTransform(counts)
  tcounts <- counts^alpha

  # collect all subsampling results
  all_res <- lapply(1:B, function(aboot) {
    selected <- sample(1:nobs, round(0.8 * nobs), replace = FALSE)
    # network inference
    res_boot <- GLMnetwork(tcounts[selected, ], lambdas = lambdas)
    return(res_boot)
  })
  # order by lambda and create adjacency
  all_res <- lapply(seq_along(lambdas), function(indl) {
    lapply(all_res, function(alist) alist[[indl]] != 0)
  })

  # compute variabilities
  all_variabilities <- lapply(all_res, function(alist) {
    variability <- Reduce("+", alist) / B
    diag(variability) <- 0
    variability <- 4 * sum(variability * (1 - variability))
    variability <- variability / (nvar * (nvar - 1))
  }) %>% unlist()
  all_variabilities <- sapply(seq_along(all_variabilities), function(ind) {
    max(all_variabilities[1:ind])
  })

  selected <- all_variabilities < 0.05
  sel_ind <- which.max(all_variabilities[selected])

  res <- list("lambdas" = lambdas, "B" = B, "best" = sel_ind,
              "variabilities" = all_variabilities)
  class(res) <- "stars"
  return(res)
}

summary.stars <- function(object, ...) {
  print(object)
}

#' @export
#' @rdname stars
print.stars <- function(x, ...) {
  cat("Best lambda found with", x$B, "bootstrap simulations is",
      x$lambdas[x$best], "\n",
      "(variability:", x$variabilities[x$best], ")")
}

#' @import graphics
#' @export
#' @rdname stars
plot.stars <- function(x, ...) {
  args <- list(...)
  args$x <- x$lambdas
  args$y <- x$variabilities

  if (is.null(args$xlab)) args$xlab <- expression(lambda)
  if (is.null(args$ylab)) args$ylab <- "stars variability"

  do.call(plot, args)
  points(x$lambdas[x$best], x$variabilities[x$best], pch = "+", cex = 1.5,
         col = "red")
}


R/user_guide.R
## User guide launcher
#' @title View RNAseqNet User's Guide
#' @name RNAseqNetUsersGuide
#' @export
#'
#' @importFrom utils browseURL
#'
#' @description Find the location of the RNAseqNet User's Guide and optionnaly
#' opens it
#' @param html logical. Should the document returned by the function be the
#' compiled PDF or the Rmd source. Default to \code{TRUE}
#' @param view logical. Should the document be opened using the default HTML
#' viewer? Default to \code{html}. It has no effect if \code{html = FALSE}
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @details The function \code{vignette("RNAseqNet")} will find the short
#' RNAseqNet vignette that describes how to obtain the RNAseqNet User's Guide.
#' The User's Guide is not itself a true vignette because it is not
#' automatically generated during the package build process. However, the
#' location of the Rmarkdown source is returned by the function if
#' \code{html = FALSE}.
#' If the operating system is not Windows, then the HTML viewer used is that
#' given by \code{Sys.getenv("R_BROWSER")}. The HTML viewer can be changed using
#'  \code{Sys.setenv(R_BROWSER = )}.
#'
#' @return Character string giving the file location. If \code{html = TRUE} and
#' \code{view = TRUE}, the HTML document reader is started and the User's Guide
#' is opened in it.
#'
#' @examples
#' RNAseqNetUsersGuide(view = FALSE)
#' RNAseqNetUsersGuide(html = FALSE)
#' \dontrun{RNAseqNetUsersGuide()}

RNAseqNetUsersGuide <- function(html = TRUE, view = html) {
  if (html) {
    f <- system.file("misc", "RNAseqNetUsersGuide.html", package = "RNAseqNet")
    if (view) {
      if (.Platform$OS.type == "windows")
        shell.exec(normalizePath(f))
      else browseURL(paste0("file://", normalizePath(f)))
    }
  } else {
    f <- system.file("misc", "RNAseqNetUsersGuide.Rmd", package = "RNAseqNet")
    if (view) {
      warning("'RNAseqNetUserGuide.Rmd' can not be viewed.
              However, the location of the file is returned by the function.",
              call. = FALSE)
    }
  }
  return(f)
}


R/main.R
# network inference from RNAseq datasets with hd-MI
#
################################################################################
# hd-MI
################################################################################
#' @title Multiple hot-deck imputation and network inference from RNA-seq data.
#'
#' @description
#' \code{imputedGLMnetwork} performs a multiple hot-deck imputation and infers a
#' network for each imputed dataset with a log-linear Poisson graphical model
#' (LLGM).
#'
#' @param X n x p numeric matrix containing RNA-seq expression with missing rows
#' (numeric matrix or data frame)
#' @param Y auxiliary dataset (n' x q numeric matrix or data frame)
#' @param sigma affinity threshold for donor pool
#' @param m number of replicates in multiple imputation (integer). Default to 50
#' @param lambdas a sequence of decreasing positive numbers to control the
#' regularization (numeric vector). Default to \code{NULL}
#' @param B number of iterations for stability selection. Default to 20
#'
#' @export
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @references {Imbert, A., Valsesia, A., Le Gall, C., Armenise, C., Lefebvre,
#' G. Gourraud, P.A., Viguerie, N. and Villa-Vialaneix, N. (2018) Multiple
#' hot-deck imputation for network inference from RNA sequencing data.
#' \emph{Bioinformatics}. \doi{10.1093/bioinformatics/btx819}.}
#'
#' @details When input \code{lambdas} are null the default sequence of
#' \code{\link[glmnet]{glmnet}} for the first model (the one with the first
#' column of \code{count} as the target) is used. A common default sequence is
#' generated for all imputed datasets using this method.
#'
#' @examples
#' data(lung)
#' data(thyroid)
#' nobs <- nrow(lung)
#' miss_ind <- sample(1:nobs, round(0.2 * nobs), replace = FALSE)
#' lung[miss_ind, ] <- NA
#' lung <- na.omit(lung)
#' lambdas <- 4 * 10^(seq(0, -2, length = 10))
#' \dontrun{
#' lung_hdmi <- imputedGLMnetwork(lung, thyroid, sigma = 2, lambdas = lambdas,
#'                                m = 10, B = 5)
#' }
#'
#' @return S3 object of class \code{HDpath}: a list consisting of
#' \itemize{
#'   \item{\code{path}}{ a list of \code{m} data frames, each containing the
#'   adjacency matrix of the inferred network obtained from the corresonding
#'   imputed dataset. The regularization parameter is selected by StARS}
#'   \item{\code{efreq}}{ a numeric matrix of size p x p,  which indicates the
#'   number of times an edge has been predicted among the \code{m} inferred
#'   networks}
#' }

imputedGLMnetwork <- function(X, Y, sigma, m = 50, lambdas = NULL, B = 20) {
  # hot deck imputation
  imputed <- imputeHD(X, Y, sigma, m)

  # lambda selection
  selected_lambda <- lapply(imputed$data, function(counts) {
    test <- stabilitySelection(counts, lambdas, B)
    sel_ind <- test$best
    if (is.null(lambdas)) lambdas <<- test$lambdas
    return(sel_ind)
  }) %>% unlist()
  lambdas <- lambdas[1:max(selected_lambda)]

  inferred_adj <- mapply(function(counts, indl) {
    # network inference
    res_glmnet <- GLMnetwork(counts, lambdas)
    adjacency <- res_glmnet$path[[indl]]
    return(adjacency)
  }, imputed$data, selected_lambda, SIMPLIFY = FALSE)

  edge_distribution <- lapply(inferred_adj, function(ares) {
    edges <- (ares !=0) * t(ares != 0)
  })
  edge_distribution <- Reduce("+", edge_distribution)

  res <- list("path" = inferred_adj, "efreq" = edge_distribution)
  class(res) <- "HDpath"
  return(res)
}

## Methods for objects of class HDpath
#' @title Methods for 'HDpath' objects.
#' @name HDpath
#' @export
#'
#' @aliases summary.HDpath
#' @aliases print.HDpath
#' @aliases HDpath-class
#'
#' @description Methods for the result of \code{\link{imputedGLMnetwork}}
#' (\code{HDpath} object)
#' @param object \code{HDpath} object
#' @param x \code{HDpath} object
#' @param ... not used
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @examples
#' data(lung)
#' data(thyroid)
#' nobs <- nrow(lung)
#' miss_ind <- sample(1:nobs, round(0.2 * nobs), replace = FALSE)
#' lung[miss_ind, ] <- NA
#' lung <- na.omit(lung)
#' lambdas <- 4 * 10^(seq(0, -2, length = 10))
#' \dontrun{
#' lung_hdmi <- imputedGLMnetwork(lung, thyroid, sigma = 2, lambdas = lambdas,
#'                                m = 10, B = 5)
#' plot(lung_hdmi)
#' }
#'
#' @seealso \code{\link{imputedGLMnetwork}}

summary.HDpath <- function(object, ...) {
  print(object)
}

#' @export
#' @rdname HDpath

print.HDpath <- function(x, ...) {
  cat("Object of class 'HDpath'...\n",
      length(x$path), "predicted sets of coefficients from Poisson GLM after HD imputation.")
}

#' @import ggplot2
#' @importFrom ggplot2 ggplot
#' @importFrom ggplot2 aes
#' @importFrom ggplot2 geom_histogram
#' @importFrom ggplot2 theme_bw
#' @importFrom ggplot2 ggtitle
#' @importFrom ggplot2 theme
#' @importFrom ggplot2 ylab
#' @importFrom ggplot2 element_text
#' @export
#' @rdname HDpath

plot.HDpath <- function(x, ...) {
  df <- data.frame("Frequency" = x$efreq[upper.tri(x$efreq)])
  p <- ggplot(data = df, aes_string(x = "Frequency")) + geom_histogram() +
    theme_bw() + ggtitle("Distribution of edge frequency") +
    ylab("count") +
    theme(title = element_text(size = 10))

  return(p)
}

################################################################################
# convert the result of imputedGLMnetwork into an 'igraph' object
################################################################################
#' @import igraph
#' @importFrom igraph graph_from_adjacency_matrix
#' @importFrom igraph simplify
#' @importFrom methods is
#'
#' @title Convert the result of imputedGLMnetwork or a matrix into a network.
#' @export
#'
#' @description
#' \code{GLMnetToGraph} combines the m inferred networks, obtained from m
#' imputed datasets, into a single stable network or convert a matrix of
#' coefficients of a GLM model into a network (non zero coefficients are
#' converted to edges)
#'
#' @param object an object of class \code{HDpath} as obtained from the function
#' \code{\link{imputedGLMnetwork}} or a squared matrix with zero and non zero
#' values
#' @param threshold the percentage of times, among the m imputed networks, that
#' an edge has to be predicted to be in the final network. Used only for objects
#' of class \code{HDpath}. Default to 0.9
#'
#' @author {Alyssa Imbert, \email{alyssa.imbert@gmail.com}
#'
#' Nathalie Vialaneix, \email{nathalie.vialaneix@inrae.fr}}
#'
#' @references {Imbert, A., Valsesia, A., Le Gall, C., Armenise, C., Lefebvre,
#' G. Gourraud, P.A., Viguerie, N. and Villa-Vialaneix, N. (2018) Multiple
#' hot-deck imputation for network inference from RNA sequencing data.
#' \emph{Bioinformatics}. \doi{10.1093/bioinformatics/btx819}.}
#'
#' @seealso \code{\link{imputedGLMnetwork}}, \code{\link[igraph]{igraph}}
#'
#' @examples
#' data(lung)
#' data(thyroid)
#' nobs <- nrow(lung)
#' miss_ind <- sample(1:nobs, round(0.2 * nobs), replace = FALSE)
#' lung[miss_ind, ] <- NA
#' lung <- na.omit(lung)
#' lambdas <- 4 * 10^(seq(0, -2, length = 10))
#' \dontrun{
#' lung_hdmi <- imputedGLMnetwork(lung, thyroid, sigma = 2, lambdas = lambdas,
#'                                m = 10, B = 5)
#' lung_net <- GLMnetToGraph(lung_hdmi, 0.75)
#' lung_net
#' plot(lung_net)
#' }
#'
#' @return an 'igraph' object. See \code{\link[igraph]{igraph}}

GLMnetToGraph <- function(object, threshold = 0.9) {
  if (is(object, "HDpath")) {
    if (threshold < 0 | threshold > 1) {
      stop("'threshold' must be a ratio (between 0 and 1)")
    }
    adj_sel <- object$efreq > round(length(object$path) * threshold)
    net <- graph_from_adjacency_matrix(adj_sel, mode = "undirected")
    net <- simplify(net)
  } else {
    adj_sel <- object != 0
    net <- graph_from_adjacency_matrix(adj_sel, mode = "undirected")
    net <- simplify(net)
  }
  return(net)
}
